Split GitHub Actions

XhmikosR opened the pull request:
This is mostly an experiment. If it's faster (even without caching) we could do things this way since Actions run in parallel.

TODO:

- [x] Fix caches
- [ ] Adapt master branch protection rules
- [ ] Backport it to v4?

XhmikosR committed:
Split GitHub Actions.

Everything should be faster now. Also, we can restart a failed workflow and not the whole thing like before.

XhmikosR commented:
Some notes so far:

1. we end up with duplicate (push and PR) runs for some of the tasks when one of the members makes a PR. I'm not sure if we can improve this
2. `Tests` is basically the JS Tests only, so we should rename it if we go ahead with this
3. The README.md badges, not sure which we should add
4. The branch protection rules will need to be adapted if we go ahead with this

XhmikosR commented on .github/workflows/browserstack.yml:
```diff
@@ -0,0 +1,35 @@
+
+      - name: Run BrowserStack tests
+        run: npm run js-test-cloud
+        if: github.repository == 'twbs/bootstrap'
```
Not sure if this check makes sense anymore?

MartijnCuppens replied:
Why was this added in the first place?

XhmikosR commented:
@mdo: I definitely like this approach, but I don't like the duplicate runs when we make a PR ourselves

XhmikosR commented:
The secrets are not available to other forks

On Wed, Oct 2, 2019, 20:19 Martijn Cuppens <notifications@github.com> wrote:

> *@MartijnCuppens* commented on this pull request.
> ------------------------------
>
> In .github/workflows/browserstack.yml
> <https://github.com/twbs/bootstrap/pull/29448#discussion_r330669930>:
>
> > +          node-version: "10"
> +
> +      - run: node --version
> +      - run: npm --version
> +
> +      - name: Install npm dependencies
> +        run: npm ci
> +        env:
> +          CI: true
> +
> +      - name: Run dist
> +        run: npm run dist
> +
> +      - name: Run BrowserStack tests
> +        run: npm run js-test-cloud
> +        if: github.repository == 'twbs/bootstrap'
>
> Why was this added in the first place?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/twbs/bootstrap/pull/29448?email_source=notifications&email_token=AACVLNPOZSKCMCTTG6PTE63QMTJ2FA5CNFSM4I22DON2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGVJ32A#discussion_r330669930>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACVLNJ5XNCTLZKZECDYPH3QMTJ2FANCNFSM4I22DONQ>
> .
>

XhmikosR commented:
Perhaps we should change the actions to fire up on `pull_request` only? Assuming we always use PRs, which we do.

MartijnCuppens commented:
> Perhaps we should change the actions to fire up on pull_request only? Assuming we always use PRs, which we do.

Fine by me, I assume the test rerun when we push something to the PR?

XhmikosR commented:
Not sure TBH.

Let me think more about it. I know for sure this is faster, though.

MartijnCuppens commented:
The tests rerun when we push to an existing PR with only `pull_request` (tested this with another project, https://github.com/Intracto/buildozer/pull/34)

XhmikosR commented:
Yeah but then they don't run when merged.

We could use some help, maybe I should ask on the community forum.

mdo commented:
@XhmikosR Send me anything via Slack DM if I can ask internally.

XhmikosR commented:
So, I renamed "Tests" to "JS Tests". I just link to this in README.md.

1. we need to agree on the workflow names
2. we need to make sure we are not missing any tests
3. I'd still like to not run everything twice when one of the team members open a PR; I guess we'll live with that
4. **just before the merge** we need to adapt the master branch protection rules
5. after the merge I need to backport this in v4-dev

MartijnCuppens commented:
The names lgtm and it looks like everything we covered before also is covered in this PR.

XhmikosR commented:
Only thing I don't like is the double actions when a team member makes a
PR. We could live with it, but there's a limit in parallel actions which we
will hit. It's not the end of the world, it's just not optimal.

On Sun, Oct 27, 2019, 10:44 Martijn Cuppens <notifications@github.com>
wrote:

> The names lgtm and it looks like everything we covered before also is
> covered in this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twbs/bootstrap/pull/29448?email_source=notifications&email_token=AACVLNO7IJTOIEJOVWRUVS3QQVBELA5CNFSM4I22DON2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECKZFHY#issuecomment-546673311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVLNNCFV2ENTVXEDNBS7DQQVBELANCNFSM4I22DONQ>
> .
>

XhmikosR commented:
Alight, I'm not sure how to skip the duplicate runs, but this is still a lot faster than what we had.

XhmikosR committed:
Add dart-sass build test

Builds the CSS with dart-sass and lists the files in dist-sass/css folder

iamandrewluca commented:
You can also limit the runs only when some specific files were modified.

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths

mdo commented:
This okay to merge @XhmikosR?

XhmikosR commented:
> This okay to merge @XhmikosR?

I'm still not 100% happy with the duplicate runs... Also, this approach is a little "dangerous" in the sense that we need to be 100% careful to not miss testing something.

Regardless, I'll rebase this later. I also added a dart sass build action.

ffoodd commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,36 @@
+        run: npm ci
+
+      - name: Test docs
+        run: npm run docs
```
Shouldn't we generate dist files before running the docs task, to ensure the build is accurate?

XhmikosR replied:
Not really since the docs tests don't depend on the dist files.

ffoodd commented:
> Yeah but then they don't run when merged.

You can restrict workflow on push events to specific branches (like `master`, or, say, `v5-dev`).

Since you're basically always working through PRs, a push event on `v5-dev` should always match a PR merge:


```sh
on:
  push:
      branches:
        - v5-dev
```

You can also target pull_request with a closed action and a `merged` argument equaling true:

```sh
on:
  pull_request:
    types:
        closed
    branches:
      - v5-dev

jobs:
  merge:
      - name: Just merged
        if: github.event_name == 'pull_request' && github.event.pull_request.action == 'closed' && github.event.pull_request.merged == true
```

I'm not quite sure if all of this work as intended, but currently trying…

XhmikosR commented:
Yeah, I know all this. The problem is that I personally find it handy to have CI run on push before I proceed with a PR.

ffoodd commented:
You'd want CI to run on any PRs, but also on push to `master`, am I right?

If so, I think this should do the job:

```yml
name: BrowserStack
on:
  push:
    paths:
      - 'js/**'
    branches:
      - master
  pull_request:
    paths:
      - 'js/**'
```

Restricting path would prevent to run if no relevant file is changed, and restricting to the `master` branch for `push` event seems to prevent the duplicate runs.

My current repo is way more quieter than this one obviously, but it seems to work fine in my case.

XhmikosR commented:
I thought about this, and I think we should only run tests on PRs, since we all create PRs to merge something.

That being said, my only concern is missing out testing some parts in case our scripts change...

But it definitely is faster.

Let me know your thoughts again and how do you think we should proceed.

XhmikosR committed:
Change when CI runs

Run it for all pull requests, the master branch, and any branches that end with `-ci`.

XhmikosR committed:
Switch to Node.js 12.x.

XhmikosR committed:
Merge branch 'master' into master-xmr-actions-split

XhmikosR commented:
@twbs/team this is ready for review. @mdo if this get approved, before the merge we need the master branch protection adapted.

If it's merged, I'll monitor things and if everything works good, I'm going to backport it.

XhmikosR committed:
Merge branch 'master' into master-xmr-actions-split
